### PR TITLE
fix(#3961): route provider and model credential reads through the thread-local profile channel

### LINKED ISSUE #3961: Resolve /api/providers + /api/models credentials from the thread-local profile channel (drop the os.environ mirror race)
## Summary

Follow-up from the #3957 fix (#3960, v0.51.356). The independent review of #3960 surfaced a **pre-existing** architectural limitation in how the read-only provider/model endpoints resolve per-profile credentials, recommending a tracking issue.

## The limitation

WebUI profile switching is per-client/cookie-scoped (#798). The fix in #3960 routes `/api/providers` and `/api/models` (and the detached models-rebuild worker) through the active profile via `profile_env_for_active_request` / `profile_scope_for_detached_worker`, which delegate to the **pre-existing** `profile_env_for_background_worker` (the same mechanism streaming uses for title-gen / checkpoints).

That mechanism uses **two channels**:
1. a **thread-local** (`_thread_ctx`) — race-free, the primary routing channel; and
2. a **process-global `os.environ` mirror**, kept because several production Hermes readers still call `os.getenv()` directly for provider credentials.

By explicit design (`api/profiles.py` ~L739-743), `_ENV_LOCK` is held only around the env **setup/restore**, NOT the `yield` body (unlike `_cron_env_lock`, which serializes the whole body). The server is a `ThreadingHTTPServer` (`daemon_threads=True`), so requests run concurrently.

**Consequence:** two clients on **different non-default profiles** hitting `/api/providers` concurrently can interleave their `os.environ` mutations during the (multi-second) serial auth-probe window — transiently bleeding one profile's credentials into the other's probe, and potentially leaving a non-default profile's value as the process-global "default" after both restore.

## Scope / severity

- **Pre-existing, not introduced by #3960.** #3960's diff to `api/profiles.py` is purely additive — `profile_env_for_background_worker` itself is unchanged. Streaming already exercises the identical `os.environ`-mirror race for concurrent different-profile title-gen/checkpoint workers.
- **`/api/models` is largely insulated**: its rebuild is serialized by `_cache_build_cv` (one rebuild at a time), and its cache path + fingerprint resolve through the **thread-local** (race-free). The exposure concentrates on `/api/providers` and the opt-in synchronous rebuild (`_LIVE_REBUILD_BUDGET_SECONDS <= 0`).
- **Same-profile concurrent clients are fine** (identical values → no bleed). The bug needs ≥2 *distinct* non-default profiles hitting `/api/providers` simultaneously, and is transient / self-correcting on the next non-concurrent load.

## Recommended fix

Resolve provider credentials for `/api/providers` + `/api/models` from the **thread-local profile channel** rather than the `os.environ` mirror — eliminating the cross-profile race for these read endpoints without process-global mutation. This spans the streaming usage too (the shared `profile_env_for_background_worker` mechanism), so it's a deliberate, broader change rather than a quick patch:

- Audit which downstream readers (`hermes_cli.auth.get_auth_status`, `hermes_cli.models.provider_model_ids`, `_lookup_custom_api_key_env`) read `os.getenv` directly vs. accept an explicit credential/home argument.
- Provide a thread-local-first resolution path so the read endpoints never need the `os.environ` mirror.
- Keep the default-profile path a no-op / unchanged.

Refs: #3957, #3960, #798.

### PR DESCRIPTION
**Pull request by rodboev**
## Thinking Path

- `#3957` and `#3960` already routed the active profile into `/api/providers` and `/api/models`, but `load_pool()` and detached model-rebuild workers could still see server-default credentials when a named profile lacked that key.
- The remaining leak was therefore not the profile cookie or cache path, it was the pool/env fallback layer underneath the readonly provider/quota reads and the mirrored background-worker rebuild path.
- The fix keeps readonly provider/quota reads fail-closed against the active profile's env plus `auth.json`, while hardening the mirrored worker scope so legacy `provider_model_ids()` and `load_pool()` readers no longer inherit absent secrets from the process default.

## What Changed

- `api/config.py`: readonly pool-backed checks now read explicit `auth.json` pool entries directly when process-env fallback is blocked, while non-readonly paths keep using the cached `load_pool()` path.
- `api/providers.py`: provider key lookup and local pool snapshots now consume that same explicit-entry path, so readonly profile reads cannot seed or detect process-default credentials through the pool layer.
- `api/profiles.py`: mirrored background-worker scopes now scrub known provider credential vars plus custom-provider `key_env` names that are absent from the active profile, and they set `block_process_env_fallback` for the duration of that mirrored scope.
- `api/routes.py`: `/api/providers` and `/api/provider/quota` stay on the readonly scope, while `/api/models/live` and mirrored rebuild readers keep their legacy path until the downstream agent helpers stop depending on direct process-env reads.
- `tests/test_issue3957_profile_providers_models.py`: coverage now includes readonly pool-seeding regression checks, detached-worker pool-seeding regression checks, custom `key_env` scrubbing, and the no-`auth.json` side effect on empty named profiles.

## Why It Matters

Named profiles should never read or persist the server's default provider credentials just because the process env has them exported. This closes that remaining bleed on `/api/providers`, `/api/provider/quota`, and the detached `/api/models` rebuild path without widening the request-level readonly scope into a process-global env mutation.

## Verification

- `pytest tests/test_issue3957_profile_providers_models.py -v --timeout=60`

Full-suite CI context, not a required local check unless requested: `pytest tests/ -v --timeout=60`.

## Risks / Follow-ups

- The mirrored request/background-worker path remains for `/api/models/live`, streaming, and other agent-side readers that still depend on process env; this PR tightens that legacy surface but does not remove it.
- Any future readonly provider/model helper that touches credential-pool state needs the same fail-closed contract: raw profile `auth.json` entries only, never a process-env-seeding `load_pool()` call.

## Upstream

Closes #3961. Follow-up to #3960.

## Model Used

GPT 5.5 via Codex CLI

### CODE REVIEW

**Comment by nesquena-hermes** [comment]
Thanks @rodboev — the direction is right (routing the remaining read-only credential surfaces through the profile-keyed thread-local path closes the last of the race surface after #3957/#3960). Suite is green (9727/0). Two narrow findings before it ships:

**1. [SILENT] `/api/models` sync rebuild can fall back to process env** (`api/config.py:6432`). The route now enters `profile_env_for_active_request_readonly()`, but `_build_available_models_uncached()` calls `hermes_cli.models.provider_model_ids()` paths that still read raw `os.getenv()` for live catalogs. With `HERMES_WEBUI_MODELS_REBUILD_BUDGET<=0`, a named profile's env-only key can be ignored or replaced by the process env. **Fix:** keep the sync rebuild on `profile_env_for_active_request()` (non-readonly) until every `provider_model_ids()` path it exercises is thread-local/profile-env aware.

**2. [SILENT] `/api/provider/quota` subprocess probe can still use process-default credentials** (`api/providers.py:1397`). The route enters the read-only thread-local scope, but `_account_usage_subprocess_env()` starts from `dict(os.environ)` and only overlays the profile `.env` — so absent profile credentials fall through to process env in the child probe. **Fix:** when `block_process_env_fallback` is set, scrub known provider credential env vars from the child env before overlaying the profile `.env` and explicit `api_key`.

Both are narrow edge cases (named-profile with env-only keys + the budget/sync path or the quota probe), so these don't affect single-profile setups or most multi-profile users. Add regression tests for each and I'll fast-track the re-review.

File: api/config.py [hunk 1]
```diff
@@ -6893,6 +6893,25 @@ def _evict_session_agent(session_id: str) -> None:
...
+    env_name = str(name or "").strip()
+    if not env_name:
+        return default or ""
+
+    thread_env = getattr(_thread_ctx, "env", {})
+    if isinstance(thread_env, dict) and env_name in thread_env:
+        thread_value = thread_env.get(env_name)
+        if thread_value is None:
+            return default
+        return str(thread_value)
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/config.py:6966]: Fixed in `7a0d8686`. The `thread_value is None` branch now returns `default or ""`, matching the rest of `_thread_local_env_value()` and keeping the helper on its string-return contract. Added `test_thread_local_env_value_none_default_returns_empty_string` to pin it.

File: api/profiles.py [hunk 2]
```diff
@@ -901,31 +901,89 @@ def profile_env_for_background_worker(
...
+            home_override_token = set_hermes_home_override(profile_home_path)
+        yield
+    finally:
+        if home_override_token is not None and reset_hermes_home_override is not None:
+            reset_hermes_home_override(home_override_token)
+        _thread_ctx.block_process_env_fallback = previous_block_process_env
+        if previous_thread_env:
+            _set_thread_env(**previous_thread_env)
+        else:
+            _clear_thread_env()
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/profiles.py:1107]: Fixed in `7a0d8686`. `profile_env_for_active_request_readonly()` now logs and continues if `reset_hermes_home_override()` raises, then still restores `block_process_env_fallback` and the previous thread-local env before the request thread returns to the pool. Added `test_active_request_scope_restores_state_when_home_reset_fails` to pin that cleanup path.

**Comment by rodboev** [comment]
@nesquena-hermes addressed both follow-ups in `7a0d8686`.

- The budget<=0 sync `/api/models` rebuild is back on `profile_env_for_active_request()`, so the foreground `provider_model_ids()` path still sees the request profile's mirrored env and Hermes-home while those raw `os.getenv()` readers remain in play.
- `_account_usage_subprocess_env()` now strips inherited provider credential env vars when the readonly request scope has `block_process_env_fallback` set, then overlays the profile `.env` and any explicit API key for the child quota probe.

Focused regression coverage for the sync rebuild path, the quota subprocess scrub, the readonly cleanup path, and the `_thread_local_env_value(..., default=None)` contract is in `tests/test_issue3957_profile_providers_models.py`. Local validation passed: `pytest tests/test_issue3957_profile_providers_models.py -v --timeout=60` (`21 passed`).

**Comment by nesquena-hermes** [comment]
Thanks @rodboev — the follow-up commit `7a0d8686` fully addresses both findings from the previous round. Reading through the changes:

## What was fixed

**Sync `get_available_models` path** (`api/config.py:6418`): the import was switched back from `profile_env_for_active_request_readonly` to `profile_env_for_active_request` for the `_LIVE_REBUILD_BUDGET_SECONDS <= 0` foreground path. That's the right call — `provider_model_ids()` in this path still has raw `os.getenv()`/`HERMES_HOME` readers that need the full mirrored env, not just thread-local isolation.

**`_account_usage_subprocess_env` credential leak** (`api/providers.py:1414-1417`): the subprocess environment now strips all provider credential vars when `block_process_env_fallback` is set:
```python
if bool(getattr(_thread_ctx, "block_process_env_fallback", False)):
    for env_name in _PROVIDER_CREDENTIAL_ENV_VARS:
        env.pop(env_name, None)
```
`_PROVIDER_CREDENTIAL_ENV_VARS` is built at module load from `_PROVIDER_ENV_VAR` + `_PROVIDER_ENV_VAR_ALIASES`, so it covers the full set of known credential names without any manual enumeration.

**`_thread_local_env_value` return value** (`api/config.py:6905`): the `return default or ""` is now consistent — a `None` thread-local value no longer returns `None` instead of the default, which would have propagated into `.strip()` callers and raised.

**Hermes-home override reset resilience** (`api/profiles.py:966-974`): `reset_hermes_home_override` call is now wrapped in a try/except so a failure on teardown doesn't leak the `block_process_env_fallback` state or the thread-local env into the next request on the same thread.

## Test coverage check

The new `test_account_usage_subprocess_env_blocks_process_default_key` test (`tests/test_issue3957_profile_providers_models.py`) is well-targeted: it sets `OPENAI_API_KEY` in process env, enters a readonly profile scope, calls `_account_usage_subprocess_env`, and asserts the key is absent from the resulting dict. That's exactly the failure mode we were worried about.

The `test_models_sync_rebuild_uses_legacy_mirrored_env` test confirms that the sync path now uses the mirrored scope, not the readonly one.

## One narrow observation (non-blocking)

`_PROVIDER_CREDENTIAL_ENV_VARS` is computed by calling `_provider_credential_env_vars()` at module import time. If a plugin or future code path adds entries to `_PROVIDER_ENV_VAR` or `_PROVIDER_ENV_VAR_ALIASES` after module load, those won't be reflected. This is fine for current usage — the dicts are module-level constants — just something to keep in mind if those maps ever become mutable.

Overall this looks merge-ready. The race surface from #3961 is closed for all three endpoints (`/api/providers`, `/api/provider/quota`, and the sync models rebuild path), and the subprocess env leak is patched.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
Ran a full security ship-gate on the rebased head (Codex + Opus + suite + a direct reproduction). The direction and the thread-local channel are right, and the earlier follow-ups landed — but the gate surfaced a **deeper cross-profile leak that this PR is meant to close but doesn't yet**, so it needs one more round before it can ship.

### 🔴 Blocker: `load_pool()` still reaches the process env, so a named profile inherits the server's provider key

`profile_env_for_active_request_readonly()` sets `block_process_env_fallback` on the thread context, but it does **not** install a secret scope. `/api/providers` (and `/api/provider/quota`) reach `agent.credential_pool.load_pool()` (via `api/config.py:4291/4295`), whose env seeding calls `agent.secret_scope.get_secret()` — and when no secret scope is installed, that falls back to real `os.environ`. So the thread-local block is bypassed for the pool path.

**Reproduced** (isolated temp `HERMES_HOME`, named profile `testprofile` with an empty `.env`, process holds `OPENROUTER_API_KEY`):

```python
profiles.set_request_profile("testprofile")
with profiles.profile_env_for_active_request_readonly("read"):
    providers._provider_has_key("openrouter")   # => True  (should be False)
```

Result: `_provider_has_key("openrouter")` returns **True** from the process key, and it even **writes the process key into the named profile's `auth.json`** (`credential_pool.openrouter[...]`). That's the exact cross-profile bleed #3961 is about, still live on `/api/providers` + `/api/provider/quota`.

### 🔴 Same gap on the `/api/models` rebuild path

`profile_env_for_background_worker()` sets the thread env but leaves the process fallback enabled and mirrors profile env without clearing absent provider-credential vars, so the catalog rebuild (`config.py:5286`/`5787` → `load_pool()`) can detect/seed default-profile credentials for a named profile missing that key.

### Fix-spec

- In `profile_env_for_active_request_readonly()`: install and reset `agent.secret_scope.set_secret_scope(thread_env)` alongside the existing Hermes-home override, so `load_pool()`'s `get_secret()` resolves from the thread env instead of `os.environ`. Add a regression: empty profile `.env` + process `OPENROUTER_API_KEY` ⇒ `_provider_has_key("openrouter") is False` and **no** profile `auth.json` is created.
- Make `profile_scope_for_detached_worker()` / `profile_env_for_background_worker()` fail closed for profile-scoped credential reads too: install/reset the secret scope with the profile env, set/restore `block_process_env_fallback`, and restore-pop known provider-credential env vars absent from the active profile (don't leave process defaults visible while using the legacy mirror). Regression: `_has_explicit_pool_credentials("openrouter")` is False under `profile_scope_for_detached_worker()` for an empty named profile.

### Lower-priority (Opus, defense-in-depth — fold in if convenient)
- `_account_usage_subprocess_env` strip set omits the AWS/Bedrock pair (`AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY`) and custom-provider dynamic env names — add them at `providers.py:_provider_credential_env_vars()`.

The mechanism is close; this is about extending the block to the `secret_scope`/`load_pool` layer. Happy to re-gate as soon as the pool path fails closed. Thanks @rodboev.

**Comment by rodboev** [comment]
@nesquena-hermes addressed in `44437f7a`.

- `api/config.py` now keeps readonly pool-backed checks on explicit `auth.json` entries when `block_process_env_fallback` is set, so `/api/providers` and `/api/provider/quota` no longer seed or detect the process-default key through `load_pool()`.
- `api/providers.py` now uses that same explicit-entry path for pool-backed key lookup and local pool snapshots, so the empty named-profile repro with process `OPENROUTER_API_KEY` now returns `False` and leaves the profile's `auth.json` untouched.
- `api/profiles.py` now strips provider credential vars, plus custom-provider `key_env` names that are absent from the active profile, from the mirrored background-worker scope and sets `block_process_env_fallback` there, so detached `/api/models` rebuilds do not inherit default-profile credentials either.
- Focused regression coverage passed locally: `pytest tests/test_issue3957_profile_providers_models.py -v --timeout=60` with `24 passed`.

**Comment by rodboev** [comment]
Thanks for the thorough ship-gate review — you're right on all three points and I
hadn't audited the `agent.secret_scope` layer at all.

The core issue is that `block_process_env_fallback` only guards the branch inside
`_pool_entry_payloads()` that routes through `hermes_cli.auth.read_credential_pool()`.
Once that branch is bypassed — either because `hermes_cli` is not the active package,
because the cache is stale, or because a different call path reaches `load_pool()`
directly (e.g. `_build_provider_detected_from_auth_store()` at line 5159 in
`config.py` has its own inline `load_pool()` calls with no guard at all) — the flag
has no effect. `load_pool()` resolves secrets through `agent.secret_scope.get_secret()`,
and without a scope installed that falls through to real `os.environ`. So the flag and
the auth.json write-through are both products of the same missing install.

Here's what the rework does:

1. **Install a secret scope in `profile_env_for_active_request_readonly()`
   (`api/profiles.py` ~line 1029).** After `_set_thread_env(**thread_env)` and
   `_thread_ctx.block_process_env_fallback = True`, call
   `agent.secret_scope.set_secret_scope(thread_env)` (wrapped in try/except for
   environments where the bundled agent isn't importable). The `thread_env` dict at
   that point is already exactly what we want: the profile's `.env` contents plus
   `HERMES_HOME`, with no process-default keys. In the `finally` block, call
   `reset_secret_scope(scope_token)` alongside the existing
   `reset_hermes_home_override` restore. This closes the `load_pool()` path for
   `/api/providers` and `/api/provider/quota`.

2. **Same install in `profile_env_for_background_worker()`
   (`api/profiles.py` ~line 927).** This is the shared implementation under both
   `profile_env_for_active_request()` and `profile_scope_for_detached_worker()`, so
   installing the scope here covers the `/api/models` rebuild path too. The restore
   goes in the existing `finally` block before the `_ENV_LOCK` section. The
   `_apply_profile_env_to_process()` call already clears absent credential keys from
   `os.environ`; the secret scope adds the same guarantee to the `load_pool()` path.

3. **Extend `_account_usage_subprocess_env`'s strip set
   (`api/providers.py` ~line 1399).** The current loop only covers
   `_PROVIDER_CREDENTIAL_ENV_VARS`, which comes from `_PROVIDER_ENV_VAR` — Bedrock
   uses IAM rather than a single API-key env var so `AWS_ACCESS_KEY_ID` and
   `AWS_SECRET_ACCESS_KEY` are not in that set. The fix adds those two explicitly, then
   also calls `_profile_secret_env_names(get_active_hermes_home())` to pick up any
   custom `key_env` names from the profile's `config.yaml`. That helper already
   iterates both the built-in credential vars and the config-declared dynamic names, so
   it avoids duplicating the scan.

On the test side: `test_active_request_readonly_scope_blocks_pool_env_seed` (line 326)
and `test_detached_worker_scope_blocks_pool_env_seed` (line 517) already cover the
maintainer's exact repro scenario, but they probably pass in CI today because the
`block_process_env_fallback` guard routes them through `hermes_cli.auth.read_
credential_pool()` rather than through `load_pool()`. The rework makes those tests
correct by design rather than by test-environment luck. I'll also add:

- A test that stubs `agent.secret_scope` and verifies the scope dict passed to
  `set_secret_scope()` during `profile_env_for_active_request_readonly()` does not
  contain the process-default key (covers the install itself, not just the observable
  result).
- The same for `profile_scope_for_detached_worker()`.
- A test that `_account_usage_subprocess_env` strips `AWS_ACCESS_KEY_ID`/
  `AWS_SECRET_ACCESS_KEY` when `block_process_env_fallback` is set.
- A test for the custom `key_env` strip case.

Nothing else in the PR changes. The `block_process_env_fallback` guard in
`_pool_entry_payloads()` stays as an additional layer; the secret scope is purely
additive on top of it.

**Comment by nesquena-hermes** [comment]
Re-gated at HEAD `1dbfddd5` on a read-only worktree vs `origin/master`. All three findings from the 22:58 ship-gate are closed. Walking each:

## 🔴 → ✅ Blocker 1: `load_pool()` reaching process env on `/api/providers` + `/api/provider/quota`

`profile_env_for_active_request_readonly()` now installs the agent secret scope alongside the existing thread-env + Hermes-home override (`api/profiles.py:1063-1064`):

```python
if _secret_scope_mod is not None:
    _scope_token = _secret_scope_mod.set_secret_scope(thread_env)
    _has_scope = True
```

This is the right layer. Checking the agent contract: `agent/secret_scope.py:123` `get_secret()` reads the installed scope authoritatively and does **not** fall through to `os.environ` when a scope is present. Since `agent.credential_pool.load_pool()` seeds via `get_secret()`, installing the scope closes *every* `load_pool()` path at once — not just the ones gated by `block_process_env_fallback`. That covers the inline `load_pool()` calls at `config.py:5164` and `:5667` that had no per-call guard. The belt-and-suspenders `_pool_entry_payloads()` raw-`auth.json` branch (`config.py:4166-4189`) is a good second layer for the case where the scope module isn't importable.

`test_active_request_readonly_scope_blocks_pool_env_seed` (test file:326) pins my exact repro — empty `work` profile + process `OPENROUTER_API_KEY` ⇒ `_provider_has_key("openrouter") is False`, `_get_provider_api_key` is None, and crucially `auth.json` is **not** created. That last assertion is the cross-profile write-through I was most worried about.

## 🔴 → ✅ Blocker 2: `/api/models` rebuild / detached-worker path

`profile_env_for_background_worker()` installs the same scope (`profiles.py:938-939`), and `_account_usage_subprocess_env` now strips absent profile credential vars under `block_process_env_fallback`. Covered by `test_detached_worker_scope_blocks_pool_env_seed` (:517), `test_detached_worker_scope_installs_secret_scope` (:621), and `test_detached_worker_scope_scrubs_absent_custom_provider_key_env` (:535).

## 🟡 → ✅ Defense-in-depth: AWS/Bedrock + custom `key_env` strip

`providers.py:1399-1406` now extends the strip set with the Bedrock IAM pair plus `_profile_secret_env_names(...)` for config-declared dynamic names:

```python
_strip.update({"AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY"})
_strip.update(_profile_secret_env_names(get_active_hermes_home()))
```

Covered by `test_account_usage_subprocess_env_strips_bedrock_keys` (:674) and `..._strips_custom_key_env` (:697).

## Note on the import hardening (commits `466412a7`, `1dbfddd5`)

Switching from a bare `from agent.secret_scope import ...` + manual `sys.modules.pop('agent', None)` to the `find_spec` probe is the correct call — popping `'agent'` wholesale could have evicted a legitimately-loaded sibling agent module mid-request. The `find_spec`-then-`__import__(fromlist=...)` shape avoids leaving a partial entry in `sys.modules` without touching unrelated submodules. Good cleanup.

## Verdict

The mechanism is now correct by design rather than by test-environment luck (your own phrasing, and accurate — the prior tests passed via the `hermes_cli.auth` branch, not `load_pool`). I'd run the focused suite once more on CI to confirm the scope-install tests are green on a shard where `agent.secret_scope` *is* importable (so the install path actually executes rather than the `_secret_scope_mod is None` fallback): `pytest tests/test_issue3957_profile_providers_models.py -v`. Assuming that's green, this is ship-ready from the security side. Nice work threading the scope through the whole credential surface.

**Comment by nesquena-hermes** [comment]
Shipped in **v0.51.590** (Release UW) 🚀 — thank you @rodboev.

Your thread-local profile credential channel is the core of this fix and it closed the #3961 leak on `/api/providers`, `/api/provider/quota`, and the sync `/api/models` rebuild.

During the release gate, Codex reproduced several *additional* cross-profile credential vectors on the same boundary — all pre-existing (env vars the agent already reads, not introduced by your PR). Rather than bounce, I extended the fix on your branch (attribution preserved via `Co-authored-by` trailers) so the whole class is closed in one ship:

- agent auth-registry OAuth tokens (`ANTHROPIC_TOKEN` / `CLAUDE_CODE_OAUTH_TOKEN`), with a fail-closed floor so they're scrubbed even when the agent package isn't importable
- the generic `CUSTOM_API_KEY` + the full AWS/Bedrock credential chain (incl. container RELATIVE+FULL URI, web-identity, bearer token)
- the Azure identity + managed-identity (MSI/IMDS) families
- `${VAR}` config-template expansion routed through the profile-aware lookup so a `api_key: ${ANTHROPIC_TOKEN}` can't reconstruct the process token under a scoped read

Region/base-URL config values are deliberately preserved. Gate: full suite green (10117), Codex closed all vectors, Opus CLEAN (it also caught two issues the config hardening introduced — an import-order NameError and a reload-cache edge — both fixed + verified). Closes #3961. Appreciate the solid foundation.
